website/blog/2025-02-17-velox-primer-part-1.mdx

```diff
@@ -0,0 +1,135 @@
+stages, and present Velox concepts such as Tasks, Splits, Pipelines, Drivers,
+and Operators. 
+
+## Distributed Query Execution
+
+Velox is a library that provides the functions that go into a *query fragment* in
+a distributed compute engine. Distributed compute engines, like Presto and
+Spark run so-called *exchange parallel plans*. Exchange is also known as a data
+shuffle, and allows data to flow from one stage to the next. Query fragments
+are the things connected by shuffles, and comprise the processing that is
+executed within a single worker node. A shuffle takes input from a set of
+fragments and routes rows of input to particular consumers based on a
+characteristic of the data, i.e., a partitioning key. The consumers of the
+shuffle read from the shuffle and get rows of data from whatever producer, such
+that the partitioning key of these rows matches the consumer.
+
+Take the following query as an example:
+
+```Shell
+SELECT key, count(*) FROM T GROUP BY key
+```
+
+Suppose it has **n** leaf fragments that scan different parts of **T** (the
+green circles at stage 1). At the end of the leaf fragment, assume there is a
+shuffle that shuffles the rows based on key. There are **m** consumer fragments
+(the yellow circles at stage 2), where each gets a non-overlapping selection of
+rows based on column **key**. Each consumer then constructs a hash table keyed
+on **key**, where they store a count of how many times the specific value of
+**key** has been seen.
+
+<figure>
+    <img src="/img/velox-primer-part1.png" height= "100%" width="100%"/>
+</figure>
+
+Now, if there are 100B different values of **key**, the hash table would not
+conveniently fit on a single machine. For efficiency, there is a point in
+dividing this into 100 hash tables of 1B entries each. This is the point of
+exchange parallel scale-out. Think of shuffle as a way for consumers to each
+consume their own distinct slice of a large stream produced by multiple
+workers.
+
+Distributed query engines like Presto and Spark both fit the above description.
+The difference is in, among other things, how they do the shuffle, but we will
+get back to that later.
+
+## Tasks and Splits
+
+Within a worker node, the Velox representation of a query fragment is called a
+Task (`velox::exec::Task`). A task is informed by mainly two things: 
+
+* **Plan** - `velox::core::PlanNode`: specifies what the Task does.
+* **Splits** - `velox::exec::Split`: specifies the data the Task operates on.
+
+The Splits correspond to the plan that the Task is executing. For the first
+stage Tasks (table scanning), the Splits specify pieces of files to scan. For
+the second stage Tasks (group by), their Splits identify the table scan Tasks
+from which the group by reads its input. There are file splits
+(`velox::connector::ConnectorSplit`) and remote splits
+(`velox::exec::RemoteSplit`). The first identifies data to read, the second
+identifies a running Task.
```

Comment:
May I ask a naive question, so the `RemoteSplit` is only for input "task" which would used by exchange?

Reply:
Yes, you would only add this specific type of split if the task in hand is supposed to read data from a shuffle. If you're reading it directly from a file (table scan) you would use a regular ConnectorSplit.